### PR TITLE
plugin lettuce testcase

### PR DESCRIPTION
**Pull request by zhaoyuguang**
Please fill up those following information before send Pull request 

Pull request NO: 2152 (https://github.com/apache/incubator-skywalking/pull/2152)
Agent repository URL: https://github.com/zhaoyuguang/incubator-skywalking
Testcases repository URL: https://github.com/zhaoyuguang/skywalking-agent-testcases

### CODE REVIEW

**Comment by ascrutae** [comment]
@zhaoyuguang I think this example is quit simple. I  think you should add test code about `io.lettuce.core.protocol.AsyncCommand#onComplete` method and check if the trace is broke.

**Comment by zhaoyuguang** [comment]
> @zhaoyuguang I think this example is quit simple. I think you should add test code about `io.lettuce.core.protocol.AsyncCommand#onComplete` method and check if the trace is broke.

ok

File: lettuce-scenario/src/main/java/test/apache/skywalking/testcase/lettuce/controller/LettuceController.java [hunk 1]
```diff
@@ -40,18 +48,38 @@
...
         RedisClient redisClient = RedisClient.create("redis://" + address + ":6379");
         StatefulRedisConnection<String, String> connection0 = redisClient.connect();
         RedisCommands<String, String> syncCommand = connection0.sync();
         syncCommand.get("key");
 
 
+        RedisAsyncCommands<String, String> asyncCommands0 = connection0.async();
+        AsyncCommand<String, String, String> future = (AsyncCommand<String, String, String>) asyncCommands0.set("key_a", "value_a");
+
+        future.onComplete(s -> {
```
Comment (wu-sheng) [inline lettuce-scenario/src/main/java/test/apache/skywalking/testcase/lettuce/controller/LettuceController.java]: If this is async call, the function inside should be run in another thread, then the trace should break if you didn't do anything in plugin.

**Comment by zhaoyuguang** [comment]
Bizarre place, after web started, no problem for the first time, second time as @ascrutae said.
Can be reproduced stably.
image
image
image
image
This may be the reason why the testcase passed.

File: lettuce-scenario/src/main/java/test/apache/skywalking/testcase/lettuce/controller/LettuceController.java (hunk 1 above)
Comment (zhaoyuguang) [reply lettuce-scenario/src/main/java/test/apache/skywalking/testcase/lettuce/controller/LettuceController.java]: I think so.

**Comment by zhaoyuguang** [comment]
In another case the trace never breaks. and traceId is same one.
image
image
image

**Comment by wu-sheng** [comment]
So, still in some cases, the trace could be broken, right?

**Comment by zhaoyuguang** [comment]
Well, it's really a problem, indeed,need to use ContextManager.capture() to make a local cross-thread, and use ContextManager.continued(snapshot) at AsyncCommand#onComplete before

**Comment by zhaoyuguang** [comment]
Maybe because the debug mode is used, the whole process becomes synchronized.

**Comment by wu-sheng** [comment]
> debug mode is used

Yes, that is possible. Anyway, if it could be across thread, please use capture/continues, it is harmless and very low cost.

**Comment by zhaoyuguang** [comment]
Because Consumer#accept uses asynchronous threads, it is necessary to rewrite java.util.function.Consumer to pass ContextSnapshot across threads, but Consumer is a class after 1.8, but the sniffer compiled version is 1.6.
image
image
image

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**